### DISCUSSION TITLE
Companions with a boolean().withDefault property (encountering a type error)

### DISCUSSION THREAD

**Discussion by emallay** [simolus3/drift#1329] [Q&A] [answered]
Novice question here... I have a boolean field in my table that has a default value of 'false'. When trying to set up a companion object (to do an insert) I'm encountering a type error when trying to assign to the boolean property in the companion's constructor. I've been searching for any examples of inserts involving a ```boolean().withDefault``` but haven't found any so far. 

Here's a simplified example of the setup:

```
class Items extends Table {
  IntColumn get id => integer().autoIncrement()();
  BoolColumn get isActive => boolean().withDefault(const Constant(false))();
  TextColumn get title => text()();
}

@UseMoor(tables: [Items], daos: [ItemsDao])
class DbTest extends _$DbTest {
  static const dbFile = 'local.sqlite';
  DbTest() : super(openConnection(dbFile));

  @override
  int get schemaVersion => 1;
}

@UseDao(tables: [Items])
class ItemsDao extends DatabaseAccessor<DbTest> with _$ItemsDaoMixin {
  final DbTest db;

  ItemsDao(this.db) : super(db);

  Future<int> insertItem(ItemsCompanion item) => into(items).insert(item);
}
```

Elsewhere, I'm trying to instantiate a companion as follows:
```
  final newItem =  ItemsCompanion.insert(title: 'Testing', isActive: false);
```

However, the attempt to assign to 'isActive' is causing an error: ```The argument type 'bool' can't be assigned to the parameter type 'Value<bool>'.```

From looking at moor's internal code, I have a rough idea of how ```Value<T>``` is being used internally, and why, but I'm not clear on how my code should be interacting with the companion. Can anyone point me towards some example code showing how this should be handled and/or give me some direction on how to properly interact with ```Value<T>``` objects (in particular, ```Value<bool>```  in moor. Thanks!

**Comment by simolus3** [comment, marked as answer]
Hi! Moor introduces the `Value` wrapper to represent that a value may be absent. For a motivation, let's look at the `isActive` column and an insert. For an insert, you can do any of the following for the `isActive`column:

- You may want to set the column to `true` when inserting a new row
- You may want to set the column to `false` when inserting a new row
- You may want to use the default value (regardless of what it actually is) when inserting a new row

That's three states and a boolean only gives us two, so we're introducing a wrapper that allows us to represent all of them:

- Set to `true`: `Value(true)`
- Set to `false`: `Value(false)`
- Ignore and use default: `Value.absent()`

So long story short, you should be using

```dart
  final newItem =  ItemsCompanion.insert(title: 'Testing', isActive: Value(false));
```

If you want to use the default for `isActive`, the easiest way is to simply use this instead:

```dart
  final newItem =  ItemsCompanion.insert(title: 'Testing');
```

I hope this helps!

-----

I can see how in this case we could simply be generating a `bool?` parameter instead and avoid the `Value` wrapper. A `null` value would mean that the default value is used, and an non-null value would override it. It only works for non-nullable columns though, since nullable columns have a fourth "I want to set the column to `null` when inserting a new row" option. So the `Value` wrapper is the most consistent approach here describing whether a value should be part of the insert or not.

**Reply by emallay** [reply]
That's a great answer and is crystal clear. Thank you very much for taking the time.
